src/main/java/uk/gov/ons/ctp/response/action/service/ActionProcessingService.java

```diff
@@ -35,66 +33,119 @@
+  }
 
+  private boolean validate(final Action action) {
     final ActionType actionType = action.getActionType();
     if (!valid(actionType)) {
       log.with("action", action).error("ActionType is not defined for action");
-      throw new IllegalStateException(ACTION_TYPE_NOT_DEFINED);
+      return false;
+    } else {
+      return true;
     }
+  }
 
-    List<ActionRequest> actionRequests = prepareActionRequests(action);
-
-    ActionDTO.ActionEvent event =
-        actionType.getResponseRequired()
-            ? ActionDTO.ActionEvent.REQUEST_DISTRIBUTED
-            : ActionDTO.ActionEvent.REQUEST_COMPLETED;
-    transitionAction(action, event);
-
-    actionRequests.forEach(
-        actionRequest -> {
-          if (actionRequest.isPubsub()) {
-            log.with("actionId", actionId.toString())
-                .info("Pubsub message will be forwarded to notify cloudfunction");
+  public void processEmails(ActionType actionType, List<Action> allActions) {
+    log.with("actions", allActions.size()).debug("processing emails");
+    for (Action action : allActions) {
+      if (checkCaseExists(action) && validate(action) && !cancelled(action)) {
+        // send to pub sub
+        List<ActionRequest> actionRequests = prepareActionRequests(action);
+        try {
+          for (ActionRequest actionRequest : actionRequests) {
             notifyService.processNotification(actionRequest);
-          } else {
-            actionInstructionPublisher.sendActionInstruction(
-                actionType.getHandler(), actionRequest);
           }
-        });
+          // if successful transition action
+          transitionAction(action, actionType);
+        } catch (RuntimeException e) {
+          log.with("action id", action.getId()).error("Error sending email", e);
+          // action is not transitions
+        }
+      }
+    }
+  }
+
+  private boolean cancelled(final Action action) {
+    if (action.getState() == ActionDTO.ActionState.CANCEL_SUBMITTED) {
+      processActionCancel(action);
+      return true;
+    } else {
+      return false;
+    }
   }
 
-  private List<ActionRequest> prepareActionRequests(Action action) {
-    final ActionRequestContextFactory factory = getActionRequestDecoratorContextFactory();
-    final ActionRequestContext context = factory.getActionRequestDecoratorContext(action);
+  private boolean checkCaseExists(final Action action) {
+    ActionCase actionCase = actionCaseRepo.findById(action.getCaseId());
+    if (actionCase == null) {
+      log.with("action", action).info("Case no longer exists for action");
+      try {
+        actionStateService.transitionAction(action, ActionDTO.ActionEvent.REQUEST_CANCELLED);
+      } catch (CTPException ctpExeption) {
+        throw new IllegalStateException(ctpExeption);
+      }
+      return false;
+    }
+    return true;
+  }
+
+  public List<ActionRequest> prepareActionRequests(Action action) {
```

Comment:
I think this would be a good place for a javadoc saying what an actionRequest IS.

Reply:
Agreed 👍